vendor: update everything

vrothberg opened the pull request:
* If possible, update each dependency to the latest available version.

* Use releases over commit IDs and avoid vendoring branches.

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

vrothberg commented:
We should be more careful in the future. There are quite some cross-dependencies and also common dependencies with different version constraints.

vrothberg commented:
@TomSweeneyRedHat @runcom @mtrmac I think we need something similar for Skopeo and Buildah until we can transition to go modules.

mheon commented:
+300k -50k... Go honestly produces absurd diffstats.

LGTM. May the Go maintainers give us a better dependency management solution in the future.

mheon commented:
Oh god what pulled in all of hcsshim

rhatdan commented:
I think this should wait until after 1.0.

vrothberg commented:
> I think this should wait until after 1.0.

@mheon, when will we create the release-1.0 branch?

mheon commented:
@vrothberg I'm thinking Thursday morning or Wednesday afternoon

mheon commented:
(Want to get as many 1.0 marked PRs in as possible first)

vrothberg commented:
Perfect!

vrothberg commented:
@rh-atomic-bot retest

vrothberg commented:
/retest

rhatdan commented:
bot, retest this please

vrothberg commented:
Updated again as `go-criu` has releases now as well (thanks @adrianreber!). I also added a `make vendor` target, which we should use for future updates to the vendor.conf. `make vendor` whitelists `varlink/go` to avoid it from being removed every time.

mtrmac commented:
> * Use releases over commit IDs and avoid vendoring branches.

The downside of that is that git tags can be moved. Ideally I guess we would want to record both.

> @TomSweeneyRedHat @runcom @mtrmac I think we need something similar for Skopeo and Buildah until we can transition to go modules.

… yes? Yes, we clearly do, if only to pick up security fixes. If you want to port the results of your analysis to `skopeo`, I’d be happy to merge that - once.

Longer term, this is a tedious chore that I’d really like not to have to manually remember to do, and manually analyze all the packages every time.

In the Old Days, a “Linux distribution” (I know, weird) would keep up with the latest versions of all the dependencies and reconcile API drift.

One benefit of `vendor.conf` referring to `master` is that every time someone runs `make vendor`, the latest updates get pulled int. The downside is that it’s non-trivial to track down which updates got pulled in, but they did actually get pulled, instead of remaining frozen at the `vendor.conf`-specified release. (OTOH, IIRC both `buildah` and `podman` got the worst of both words, by using `master` references but not keeping up with the API changes, so everyone just vendored in the one dependency that was necessary and `vendor.conf` drifted away from the actual `vendor` tree contents.)

I guess we could make “update vendor.conf” an expected part of each release process; that would still be manual and tedious, but at least a regular process.

Are there any tools to help with this, e.g. looking up the latest tag from each repo? I think there must be…

I’m also very tempted to defer worrying about this at all (or, to let the Linux distributions worry about it… I know, that does not actually shift the work at all) until we port everything Go modules.  At that point, at least, we could have a shared “common containers/* dependencies” module which would require the latest versions of every dependency, so that the other repositories did not need redundantly updating all the time.

vrothberg commented:
> > * Use releases over commit IDs and avoid vendoring branches.
> 
> The downside of that is that git tags can be moved. Ideally I guess we would want to record both.

That's an extreme but valid example (see github.com/moby/moby). We could record the commti IDs in comments above the specific line in the`vendor.conf`.

> > @TomSweeneyRedHat @runcom @mtrmac I think we need something similar for Skopeo and Buildah until we can transition to go modules.
> 
> … yes? Yes, we clearly do, if only to pick up security fixes. If you want to port the results of your analysis to `skopeo`, I’d be happy to merge that - once.

Getting that into Skopeo and Buildah should be straight forward (please don't quote me :) once everything's cleaned up here.

> Longer term, this is a tedious chore that I’d really like not to have to manually remember to do, and manually analyze all the packages every time.
> 
> In the Old Days, a “Linux distribution” (I know, weird) would keep up with the latest versions of all the dependencies and reconcile API drift.
> 
> One benefit of `vendor.conf` referring to `master` is that every time someone runs `make vendor`, the latest updates get pulled int. The downside is that it’s non-trivial to track down which updates got pulled in, but they did actually get pulled, instead of remaining frozen at the `vendor.conf`-specified release. (OTOH, IIRC both `buildah` and `podman` got the worst of both words, by using `master` references but not keeping up with the API changes, so everyone just vendored in the one dependency that was necessary and `vendor.conf` drifted away from the actual `vendor` tree contents.)

I share your opinion as it's more convenient for development. However, we love to increase our user base by getting the tools into more distributions, Debian in particular. In this case, we don't have much choice and need to be explicit  about the dependencies.

> I guess we could make “update vendor.conf” an expected part of each release process; that would still be manual and tedious, but at least a regular process.

Wonderful idea!

> Are there any tools to help with this, e.g. looking up the latest tag from each repo? I think there must be…

I don't know but would love to have something that could semi-automate this task.

> I’m also very tempted to defer worrying about this at all (or, to let the Linux distributions worry about it… I know, that does not actually shift the work at all) until we port everything Go modules. At that point, at least, we could have a shared “common containers/* dependencies” module which would require the latest versions of every dependency, so that the other repositories did not need redundantly updating all the time.

I think that Go modules will ease much of our pain but it doesn't solve breaking builds when APIs change and dependencies don't update (e.g., https://github.com/openshift/imagebuilder/issues/116). We are currently working with some Debian folks on packaging and I really think we should make packaging the tools as easy as possible. Otherwise, it is a very frustating task to untangle the dependencies and discovering those breaking  builds.

mtrmac commented:
> We could record the commti IDs in comments above the specific line in the`vendor.conf`.

That would realistically only work reliably when automated. (IIRC some vendoring tools do, or used to do, this. Basically treat my comments as a list of feature requests for some indeterminate future.)

> Debian in particular. In this case, we don't have much choice and need to be explicit about the dependencies.

Ah, OK, I didn’t realize that is the motivation.

> I think that Go modules will ease much of our pain but it doesn't solve breaking builds when APIs change and dependencies don't update (e.g., [openshift/imagebuilder#116](https://github.com/openshift/imagebuilder/issues/116)).

In theory Go modules make it possible to vendor different major versions side by side, but that could also require very extensive work inside the packages to make it possible (if any singletons exist) — and only if the conflicting dependency actually exists as a Go module.

mheon commented:
@vrothberg Now that 1.0 is branched, we can merge this. Mind rebasing?

vrothberg commented:
> @vrothberg Now that 1.0 is branched, we can merge this. Mind rebasing?

Thanks for the ping, @mheon! Sure, I will take care of it.

vrothberg commented:
/test images

vrothberg commented:
/approve

rhatdan commented:
/lgtm
